dts: Add extra definition for preprocessor logic

finikorg opened the pull request:
At the moment a lot of code depends on preprocessor static memory
allocations where we cannot use strings created by DTS. For example
string: MAXIMUM_SPEED="full-speed". Create also additional definition
looking like MAXIMUM_SPEED_FULL_SPEED=1 which can be used if somebody
needs it in preprocessor.

finikorg committed:
dts: Add extra definition for preprocessor logic

At the moment a lot of code depends on preprocessor static memory
allocations where we cannot use strings created by DTS. For example
string: MAXIMUM_SPEED="full-speed". Create also additional definition
looking like MAXIMUM_SPEED_FULL_SPEED=1 which can be used if somebody
needs it in preprocessor.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

ulfalizer commented on scripts/dts/extract/default.py:
```diff
@@ -64,6 +64,12 @@ def extract(self, node_path, prop, prop_type, def_label):
+            new_prop = {}
+            prop_val = label + '_' + str_to_label(prop_def[label]).strip('"')
+            new_prop[prop_val] = '1'
+            insert_defs(node_path, new_prop, {})
```
Could shorten this down to

```python
prop_val = label + '_' + str_to_label(prop_def[label]).strip('"')
insert_defs(node_path, {prop_val: '1'}, {})
```

or the like and get rid of `new_prop`. `prop_val` seems like a weird variable name for a macro name too, but maybe it's consistent with some other thing.

I want to replace `insert_defs()` with a function that just inserts a single name/value at some point. It's awkward that you have to call it with dictionaries.

ulfalizer replied:
Wonder if we could start a better tradition in the DTS code and add a comment too, mentioning what kind of stuff this generates, and why it's handy.

finikorg replied:
@ulfalizer I can add comments describing the change, I just wanted to get feedback about the proposed solution.

ulfalizer replied:
FWIW, at least GCC is clever enough to optimize away `strcmp()`s with constant arguments:

```c
#include <string.h>

#define FOO "hello"

int f(void)
{
    return strcmp(FOO, "hello") ? 0x123 : 0x456;
}
```

gives

```
   0:	b8 56 04 00 00       	mov    eax,0x456
   5:	c3                   	ret    
```

Could use `__builtin_strcmp()` to make sure GCC's builtin version is used.

If we don't want to rely on that, and don't want to have case-by-case code for string properties, then something like this PR is probably the only option.

finikorg replied:
The point was to use it in the preprocessor, can we use strcmp there?

ulfalizer replied:
> The point was to use it in the preprocessor, can we use strcmp there?

Nah, won't be as general.

Could have a zero-overhead `IS_ENABLED()`-like though.

finikorg replied:
> > The point was to use it in the preprocessor, can we use strcmp there?
> 
> Nah, won't be as general.
> 
> Could have a zero-overhead `IS_ENABLED()`-like though.

Do you mean we can use somehow strings in preprocessor?

ulfalizer replied:
@finikorg 
I mean that if you just need to conditionally run some code depending on the value of a DT string property, then it could be implemented with zero overhead by relying on the compiler to optimize away `strcmp()`s with constant arguments, similar to how `IS_ENABLED()` works for Kconfig symbols.

```c
if (DT_STREQ(MAXIMUM_SPEED, "full-speed")) {
        ...
} else if (DT_STREQ(MAXIMUM_SPEED, "yadayada")) {
        ...
```

`DT_STREQ` could use `__builtin_strcmp()` or similar, if we assume GCC/Clang.

I don't know how reliable that is with other compilers, and it's less general than your solution. Just another option.

finikorg replied:
@ulfalizer That strings should be evaluated only once so we can leave strcmp as is since it is much more readable.

ulfalizer replied:
Should just be careful so that the compiler doesn't turn it into an actual call in that case, for whatever reason ("hmm, perhaps this isn't guaranteed to be the standard `strcmp()`, due to compilation setting `-foo`, so I can't safely optimize it").

The `__builtin_` variants are nice in that they're guaranteed to avoid stuff like that.

finikorg replied:
That is probably needs to be done in `strings` or wherever `strcmp` is implemented.

finikorg replied:
> Could shorten this down to
> 
> ```python
> prop_val = label + '_' + str_to_label(prop_def[label]).strip('"')
> insert_defs(node_path, {prop_val: '1'}, {})
> ```
> 
> or the like and get rid of `new_prop`.
> 
> `prop_val` seems like a weird variable name for a macro name too, but maybe it's consistent with some other thing. Could it be called `name` or something instead?
> 
> I want to replace `insert_defs()` with a function that just inserts a single name/value at some point. It's awkward that you have to call it with dictionaries.

simplified this way

finikorg committed:
soc: stm32f7: Add fixups for USB maximum-speed parameter

Add defines DT_USB_MAXIMUM_SPEED_HIGH_SPEED and
DT_USB_MAXIMUM_SPEED_FULL_SPEED

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
soc: same70: Add fixups for USB maximum speed

Add DT_USB_MAXIMUM_SPEED_FULL_SPEED definition.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
soc: stm32f4: Add fixups for USB maximum-speed parameter

Add defines DT_USB_MAXIMUM_SPEED_HIGH_SPEED and
DT_USB_MAXIMUM_SPEED_FULL_SPEED

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
soc: stm32f2: Add fixups for USB maximum-speed parameter

Add defines DT_USB_MAXIMUM_SPEED_HIGH_SPEED and
DT_USB_MAXIMUM_SPEED_FULL_SPEED

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
soc: stm32l4: Add fixups for USB maximum-speed parameter

Add defines DT_USB_MAXIMUM_SPEED_HIGH_SPEED and
DT_USB_MAXIMUM_SPEED_FULL_SPEED

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

ulfalizer commented on scripts/dts/extract/default.py:
```diff
@@ -64,6 +64,13 @@ def extract(self, node_path, prop, prop_type, def_label):
+        # definition SPEED_HIGH_SPEED=1 which can be used in preprocessor logic
+        # where we do not have strcmp()
+        if prop_type == 'string':
+            prop_val = label + '_' + str_to_label(prop_def[label]).strip('"')
```
Could maybe make it `prop_val = label + '_' + str_to_label(prop_def[label].strip('"'))`, or have `str_to_label()` strip `"`.

ulfalizer reviewed (approved):
`scripts/dts` code looks okay to me. Don't know enough about where this is used to know if it's worthwhile otherwise.

carlescufi reviewed (approved):
I am happy with this solution

jfischer-no reviewed (approved):
LGTM

jfischer-phytec-iot commented:
What was the trick so that one do not need to add this to dts_fixup.h?

galak reviewed (changes requested):
Marked request for change until I have time to review.

finikorg commented:
> Marked request for change until I have time to review.

@galak Do you have time now to check this?

galak commented:
Sorry I haven't had time to look at this sooner.  I'd like to solve this by adding the concept of a enum to the yaml.

something like:

```
     maximum-speed:
+      type: string-enum
+      enum:
+        - super-speed
+        - high-speed
+        - full-speed
+        - low-speed
```

What I'm trying to figure out is what exactly we should generate, any suggestions on what the define might look like.

finikorg commented:
> Sorry I haven't had time to look at this sooner. I'd like to solve this by adding the concept of a enum to the yaml.
> 
> something like:
> 
> ```
>      maximum-speed:
> +      type: string-enum
> +      enum:
> +        - super-speed
> +        - high-speed
> +        - full-speed
> +        - low-speed
> ```
> 
> What I'm trying to figure out is what exactly we should generate, any suggestions on what the define might look like.

I have 2 solutions: 
 - one is to have it int and enum like 715c7b485b30178f6482fac1572bc1fe4bc45da6
 - another is to generate defines from above: if we have  maximum-speed="high-speed" then we have generated `MAXIMUM_SPEED_HIGH_SPEED=1`.

carlescufi commented:
@galak can you please follow-up this one?

galak commented:
> @galak can you please follow-up this one?

Can you try this and see if it works.  This will generate something like:
```
#define DT_ST_STM32_OTGFS_50000000_MAXIMUM_SPEED_ENUM		2
```
We need a matching enum definition somewhere.


```
diff --git a/dts/bindings/usb/usb.yaml b/dts/bindings/usb/usb.yaml
index 77fb17be48..6411a9d4e4 100644
--- a/dts/bindings/usb/usb.yaml
+++ b/dts/bindings/usb/usb.yaml
@@ -26,6 +26,11 @@ properties:
                    via DT, USB controllers should use their maximum
                    hardware capability.
       generation: define
+      enum:
+         - "super-speed"
+         - "high-speed"
+         - "full-speed"
+         - "low-speed"
 
```


```
diff --git a/scripts/dts/extract/default.py b/scripts/dts/extract/default.py
index 041013c206..eeee1b3af8 100644
--- a/scripts/dts/extract/default.py
+++ b/scripts/dts/extract/default.py
@@ -11,6 +11,21 @@ from extract.directive import DTDirective
 # @brief Manage directives in a default way.
 #
 class DTDefault(DTDirective):
+    def _extract_enum(self, node_path, prop, prop_values, label):
+        cell_yaml = get_binding(node_path)['properties'][prop]
+        if 'enum' in cell_yaml:
+            if prop_values in cell_yaml['enum']:
+                if isinstance(cell_yaml['enum'], list):
+                   value = cell_yaml['enum'].index(prop_values)
+                if isinstance(cell_yaml['enum'], dict):
+                   value = cell_yaml['enum'][prop_values]
+                label = label + "_ENUM"
+                return {label:value}
+            else:
+                print("ERROR")
+        return {}
+
+
     ##
     # @brief Extract directives in a default way
     #
@@ -49,6 +64,9 @@ class DTDefault(DTDirective):
             if prop_values == 'parent-label':
                 prop_values = find_parent_prop(node_path, 'label')
 
+            prop_def = self._extract_enum(node_path, prop, prop_values, label)
+            if prop_def:
+                add_compat_alias(node_path, prop_name + "_ENUM" , label + "_ENUM", prop_alias)
             if isinstance(prop_values, str):
                 prop_values = "\"" + prop_values + "\""
             prop_def[label] = prop_values
```

finikorg commented:
@galak I am OK with this method. Could you create a patch and I will rebase on top of it.

carlescufi commented:
No longer applicable
